File: Python (Assignment 1.2 - Object-oriented)/PatternFactory.py
from abc import abstractmethod, ABC
import PatternStyles


class PatternFactory(ABC):
    @abstractmethod
    def init(self, image, init_patterns):
        pass


class SimpleTheme(PatternFactory):
    def init(self, image, init_patterns):
        init_patterns.extend([
            PatternStyles.Stripes(image),
            PatternStyles.Waves(image),
                              ])


class ComplexTheme(PatternFactory):
    def init(self, image, init_patterns):
        init_patterns.extend([
            PatternStyles.Psychedelic(image),
        ])

File: C++ (Assignment 1.2 - Object-oriented)/PatternFactory.cpp
#include <algorithm>

#include "PatternFactory.h"
#include "PatternStyles.h"
#include "BaseImage.h"

void patternFactory::insertPatterns(std::vector<pattern*>& patternsFrom, std::vector<pattern*>* patternsTo) {
	std::copy(patternsFrom.begin(), patternsFrom.end(), std::back_inserter(*patternsTo));
}

void simplePatternsFactory::init(baseImage* image, std::vector<pattern*>* initPatterns) {
	
	std::vector<pattern*> patterns{ 
		new Stripes(image),
		new Waves(image),
	}; 

	insertPatterns(patterns, initPatterns);
}


void complexPatternsFactory::init(baseImage* image, std::vector<pattern*>* initPatterns) {

	std::vector<pattern*> patterns{
		new Psychedelic(image),
	};

	insertPatterns(patterns, initPatterns);
}





